tools/clang/unittests/HLSLExec/LongVectors.cpp

```diff
@@ -2,25 +2,235 @@
+    return CompareDoubleEpsilon(A, B, Tolerance);
+  case LongVector::ValidationType_Ulp: {
+    // Tolerance is in ULPs. Convert to int64_t for the comparison.
+    const int64_t IntTolerance = static_cast<int64_t>(Tolerance);
+    return CompareDoubleULP(A, B, IntTolerance);
+  };
+  default:
+    WEX::Logging::Log::Error(
+        L"Invalid ValidationType. Expecting Epsilon or ULP.");
+    return false;
+  }
+}
+
+template <typename DataTypeT>
+bool LongVector::doVectorsMatch(const std::vector<DataTypeT> &ActualValues,
+                                const std::vector<DataTypeT> &ExpectedValues,
+                                float Tolerance,
+                                LongVector::ValidationType ValidationType) {
+  // Stash mismatched indexes for easy failure logging later
+  std::vector<size_t> MismatchedIndexes;
+  VERIFY_IS_TRUE(ActualValues.size() == ExpectedValues.size(),
+                 L"doVectorsMatch() called with mismatched vector sizes.");
+  for (size_t i = 0; i < ActualValues.size(); ++i) {
+    if (!doValuesMatch(ActualValues[i], ExpectedValues[i], Tolerance,
+                       ValidationType))
+      MismatchedIndexes.push_back(i);
+  }
+
+  if (MismatchedIndexes.empty())
+    return true;
+
+  if (!MismatchedIndexes.empty()) {
+    for (size_t Index : MismatchedIndexes) {
+      std::wstringstream Wss(L"");
+      Wss << std::setprecision(15); // Set precision for floating point types
+      Wss << L"Mismatch at Index: " << Index;
+      Wss << L" Actual Value:" << ActualValues[Index] << ",";
+      Wss << L" Expected Value:" << ExpectedValues[Index];
+      WEX::Logging::Log::Error(Wss.str().c_str());
+    }
+  }
+
+  return false;
+}
+
+// A helper to fill the expected vector with computed values. Lets us factor out
+// the re-used std::get_if code and the for loop.
+template <typename DataTypeT, typename ComputeFnT>
+void fillExpectedVector(LongVector::VariantVector &ExpectedVector, size_t Count,
+                        ComputeFnT ComputeFn) {
+  auto *TypedExpectedValues =
+      std::get_if<std::vector<DataTypeT>>(&ExpectedVector);
+
+  VERIFY_IS_NOT_NULL(TypedExpectedValues,
+                     L"Expected vector is not of the correct type.");
+
+  for (size_t Index = 0; Index < Count; ++Index)
+    TypedExpectedValues->push_back(ComputeFn(Index));
+}
+
+template <typename DataTypeT>
+void LongVector::logLongVector(const std::vector<DataTypeT> &Values,
+                               const std::wstring &Name) {
+  WEX::Logging::Log::Comment(
+      WEX::Common::String().Format(L"LongVector Name: %s", Name.c_str()));
+
+  const size_t LoggingWidth = 40;
+
+  std::wstringstream Wss(L"");
+  Wss << L"LongVector Values: ";
+  Wss << L"[";
+  const size_t NumElements = Values.size();
+  for (size_t i = 0; i < NumElements; i++) {
+    if (i % LoggingWidth == 0 && i != 0)
+      Wss << L"\n ";
+    Wss << Values[i];
+    if (i != NumElements - 1)
+      Wss << L", ";
```

Comment:
not a review comment: I hate that there isn't a better way to express this idea in most languages. I want languages to add a new control block called `between` that only runs code between elements in a loop